### PR TITLE
Add rueidislimiter Module for Distributed Rate Limiting

### PR DESCRIPTION
**Pull request by nesty92**
The module is inspired by [GitHub's approach](https://github.blog/engineering/infrastructure/how-we-scaled-github-api-sharded-replicated-rate-limiter-redis/) to scaling rate limiting with Redis to support high-traffic APIs in a distributed system.

## Key Features

- **Token Bucket Rate Limiting**: Implements a token bucket algorithm to limit the number of requests a client can make within a specified time window.
- **Customizable Configuration**: Allows setting custom limits and time windows for each rate limiter instance.
- **Distributed Consistency**: Uses Redis as a backend to ensure consistency and persistence of rate limits across distributed systems.
- **Reset Information**: Provides helpful information about when the next allowed request can occur `RetryAt`.

### CODE REVIEW

**Comment by rueian** [comment]
Thanks @nesty92, this is cool. I think we can deliver it in the next next release (v1.0.50) since v1.0.49 already got a lot of updates.

File: rueidislimiter/limiter.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+const PlaceholderPrefix = "rueidislimiter"
+
+type rateLimiter struct {
+	client rueidis.Client
+	key    string
+	limit  int
+	window time.Duration
+}
+
+type RateLimiterOption struct {
```
Comment (proost) [inline rueidislimiter/limiter.go:39]: I think, we need expiration duration option need. 

some cases, for example video contents, become fashionable and not long after can be out of fashion. In this case, dead rate limiter still exists wasting memory.

Only growing, never shrinking data set size is not a good idea. how about adding TTL option which must be bigger than window?

File: rueidislimiter/limiter.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+	rateLimitScript = rueidis.NewLuaScript(`
+--   count a request for a client
+--   and return the current state for the client
+-- rename the inputs for clarity below
+local rate_limit_key = KEYS[1]
+local increment_amount = tonumber(ARGV[1])
+local next_expires_at = tonumber(ARGV[2])
+local current_time = tonumber(ARGV[3])
+local expires_at_key = rate_limit_key .. ":ex"
+local expires_at = tonumber(redis.call("get", expires_at_key))
```
[the remark this answers is not in the document]
Comment (nesty92) [reply rueidislimiter/limiter.go:191]: Maybe a single Redis key with BITFIELD to store both the request counter and expiration timestamp in a single 64-bit integer. We allocate:
 - The first 32 bits for the request counter, supporting up to 4,294,967,295 requests per time window.
 - The next 32 bits for the expiration timestamp, allowing us to track expirations up to the year 2106.
we reduce Redis calls and improve efficiency in rate limiting operations.

Even with those limits it should be enough for most use cases

File: rueidislimiter/limiter.go (hunk 1 above)
Comment (proost) [reply rueidislimiter/limiter.go:39]: Oh, i missed the code!

File: rueidislimiter/limiter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nesty92) [reply rueidislimiter/limiter.go:191]: I tried both approach and here is a [bench for it](https://gist.github.com/nesty92/be1a15385e027289c58b6ce9f1f36a13). Try to replicate it and let me know if a miss something

| Size    | Concurrency | SeparateKeys (Avg RPS / Latency ms) | Bitfield (Avg RPS / Latency ms) | HSET (Avg RPS / Latency ms) |
|---------|-------------|-------------------------------------|---------------------------------|-----------------------------|
| 100     | 50          | 50000.00 / 0.79                     | 33333.33 / 1.08                 | 33333.33 / 1.16             |
| 100     | 100         | 33333.33 / 1.57                     | 25000.00 / 1.93                 | 25000.00 / 2.11             |
| 100     | 200         | 25000.00 / 2.35                     | 25000.00 / 1.79                 | 25000.00 / 1.83             |
| 100     | 500         | 20000.00 / 2.22                     | 20000.00 / 2.58                 | 20000.00 / 1.86             |
| 1000    | 50          | 52631.58 / 0.83                     | 50000.00 / 0.92                 | 55555.56 / 0.78             |
| 1000    | 100         | 66666.67 / 1.24                     | 55555.56 / 1.55                 | 55555.56 / 1.53             |
| 1000    | 200         | 50000.00 / 3.19                     | 55555.56 / 2.84                 | 47619.05 / 3.15             |
| 1000    | 500         | 40000.00 / 7.56                     | 38461.54 / 8.67                 | 33333.34 / 10.37            |
| 10000   | 50          | 57142.86 / 0.79                     | 55248.62 / 0.83                 | 52910.05 / 0.86             |
| 10000   | 100         | 63694.27 / 1.41                     | 60975.61 / 1.48                 | 60240.96 / 1.51             |
| 10000   | 200         | 58139.53 / 3.13                     | 54945.05 / 3.27                 | 53475.93 / 3.41             |
| 10000   | 500         | 48309.18 / 9.05                     | 47169.81 / 9.63                 | 45871.56 / 9.81             |
| 100000  | 50          | 53561.86 / 0.85                     | 52301.26 / 0.88                 | 53792.36 / 0.85             |
| 100000  | 100         | 64061.50 / 1.40                     | 60386.47 / 1.51                 | 56947.61 / 1.62             |
| 100000  | 200         | 55959.71 / 3.30                     | 51440.33 / 3.60                 | 52548.61 / 3.51             |
| 100000  | 500         | 49236.83 / 9.43                     | 48828.12 / 9.53                 | 47438.33 / 9.81             |

a83cde09-bd30-47d7-b50e-0864d0a10df5

It seems that the current implementation is consistently outperforming the others in both RPS and latency across various sizes and concurrency levels.

File: rueidislimiter/limiter.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+		return nil, err
+	}
+	rl.key = option.Key
+	return rl, nil
+}
+
+func (l *rateLimiter) Check(ctx context.Context, identifier string) (Result, error) {
+	now := time.Now().UTC()
+	keys := []string{l.getKey(identifier)}
+	args := []string{strconv.FormatInt(now.Unix(), 10)}
```
[the remark this answers is not in the document]
Comment (nesty92) [reply rueidislimiter/limiter.go:77]: For rate-limiting policies based on RPS, Unix() is generally precise enough. It aligns directly with the Redis expireat command's use of seconds, keeping the configuration simple and efficient.

File: rueidislimiter/limiter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rueian) [reply rueidislimiter/limiter.go:175]: Good to know that the separated keys approach is the fastest one. Will doing MGET in the script make it even faster?

File: rueidislimiter/limiter.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rueian) [reply rueidislimiter/limiter.go:75]: Why? Unless there are some performance penalties, I don't see using millisecond precision as a fuss.

[the remark this answers is not in the document]
Comment (nesty92) [reply rueidislimiter/limiter.go:75]: I’ll switch to millisecond precision with UnixMilli(), keeping TTL one second after the rate window to handle clock differences, similar to GitHub’s approach.

File: rueidislimiter/limiter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nesty92) [reply rueidislimiter/limiter.go:175]: the actual solution shows higher RPS than MGET across most configurations.

c7456d02-9c44-4545-9501-f5809b30157b

I updated the [bench](https://gist.github.com/nesty92/be1a15385e027289c58b6ce9f1f36a13)

File: rueidislimiter/limiter.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+local increment_amount = tonumber(ARGV[1])
+local next_expires_at = tonumber(ARGV[2])
+local current_time = tonumber(ARGV[3])
+local expires_at_key = rate_limit_key .. ":ex"
+local expires_at = tonumber(redis.call("get", expires_at_key))
+if not expires_at or expires_at < current_time then
+  redis.call("set", rate_limit_key, 0)
+  redis.call("set", expires_at_key, next_expires_at)
+  redis.call("pexpireat", rate_limit_key, next_expires_at + 1000)
+  redis.call("pexpireat", expires_at_key, next_expires_at + 1000)
```
Comment (rueian) [inline rueidislimiter/limiter.go:185]: Can we merge these with `SET PXAT`?

File: rueidislimiter/limiter.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+		rl.client, err = rueidis.NewClient(option.ClientOption)
+	}
+	if err != nil {
+		return nil, err
+	}
+	rl.key = option.KeyPrefix
+	return rl, nil
+}
+
+func (l *rateLimiter) Check(ctx context.Context, identifier string) (Result, error) {
```
Comment (rueian) [inline rueidislimiter/limiter.go:72]: Is using AllowN(ctx, 0) to replace this a good idea? We can maintain less code by doing that.

Comment (nesty92) [reply rueidislimiter/limiter.go]: This is a great idea, not it should be more easy to follow and mantain

**Review by rueian — APPROVED** [review]

**Pull request merged**